Python/bltinmodule.c

```diff
@@ -2628,36 +2643,86 @@ zip_next(zipobject *lz)
             item = (*Py_TYPE(it)->tp_iternext)(it);
             if (item == NULL) {
                 Py_DECREF(result);
+                if (lz->tuplesize < 0) {
+                    goto check;
+                }
                 return NULL;
             }
             PyTuple_SET_ITEM(result, i, item);
         }
     }
     return result;
+check:
+    if (PyErr_Occurred() && !PyErr_ExceptionMatches(PyExc_StopIteration)) {
+        return NULL;
+    }
+    if (i) {
+        PyErr_Clear();
+        const char* plural = i == 1 ? " " : "s 1-";
+        return PyErr_Format(PyExc_ValueError,
+                            "%s() argument %d is shorter than argument%s%d",
+                            Py_TYPE(lz)->tp_name, i + 1, plural, i);
```

Comment:
When is tp_name ever not "zip"? I suppose when the user subclasses it. But at that point we don't know anything about the argument order (the subclass could have a different constructor signature).